### ISSUE TITLE
js-translation.json does not pick up strings in *.phtml files

### ISSUE THREAD

**Issue by amenk** [magento/magento2#33245]
Follow up of https://github.com/magento/magento2/issues/8508#issuecomment-861357969

### Background

Magento automatically picks up strings which should be supplied for the frontend translation via component via js-translation.json. This misses some strings, leaving hard-to-debug problems with translations to other languages.

### Summary (*)

The process of fetching strings for js-translation.json has problems in 2.3.5 - and I am pretty sure it's still a problem in the dev master.

We struggled to translate the string "items" for the knockout component "minicart".

### Steps to reproduce (*)

1. Create a new custom module.
2. Set Locale language to French (France) in Admin configurations Stores->Configurations->General->Locale Options.
3. Add French translation for ‘foo bar 123’ in Admin Panel to “barre de foo 123”
4. Keep string  in newly created module phtml file to check if string is getting translated.
5. Flush cache and check module on frontend.

**Result:** String is not getting translated with knockout js tag. It is displayed as ‘foo bar 123’ only. 

### Examples (*)

The extraction of the files happens here:

https://github.com/magento/magento2/blob/32fda1cba6ebd1bd7837d01dc34accf67469d560/app/code/Magento/Translation/Model/Js/DataProvider.php#L97-L102

The translatable string "items" is introduced here: 

https://github.com/magento/magento2/blob/32fda1cba6ebd1bd7837d01dc34accf67469d560/app/code/Magento/Checkout/view/frontend/templates/cart/minicart.phtml#L22

But the above extraction method does not count *.phtml files in - so translations like this are lost. (it eventually might still end up in the js-translation.json if the same string is used somewhere else in the theme, so such would cover up this bug)

### Proposed solution

Include \Magento\Framework\App\Utility\Files::getPhtmlFiles or something similar in the list of files - but this might have a performance impact.

Or remove the whole automatic string extraction and add a manual list of strings which should be included in js-translation.json - this would make the whole process much simpler and less magic.

--
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by amenk** [comment]
As code changes are necessary / it's a developer experience issue, I do not know how to reproduce that by the way. But from looking at the code, the issues is pretty clear.

**Comment by mrtuvn** [comment]
i think actually issue knockout not translated text correct in phtml file
From magento 2.4 i dont see file js-translation.json contain anything when compare with previous version

**Comment by engcom-Lima** [comment]
Hi @amenk,

Since this issue is originally 3 years old and were probably testedon old versions of Magento. So can you please check whether this issue still exists on 2.4-develop instance ?

If yes, please provide steps to reproduce on 2.4-develop instance.

For checking the issue thoroughly, it is very important that the issue should be reproducible. Then only we’ll be able to find a solution to the issue.

**Comment by amenk** [comment]
@engcom-Lima No, this issue is not 3 years old. We tested on 2.3.5 and it seems a general problem.
Can you provide the commits where this was fixed?

I am pretty sure the issue is still there, as in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Translation/Model/Js/DataProvider.php#L97 also .phtml files are left out from the string collection
and in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Checkout/view/frontend/templates/cart/minicart.phtml still strings are introduced in .phtml


I currently don't have the time to reproduce this on 2.4, but I hope it's not getting closed, because if it is not fixed, it will frustrate more developers in the future.

Short steps to reproduce would be:

* introduce a string like "foo bar 123" in the .phtml
* Check the JS translation: it will not be there

This is not easy to reproduce on the Magento 2.4 Dev instances created by the GitHub bot, as we don't have SSH access there.

**Comment by mrtuvn** [comment]
https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/translations/translate_theory.html
Seem knockout not support translate from phtml if you see this document. I don't see any mention in this doc

**Comment by amenk** [comment]
Interesting catch, yeah, so it is having a flaw by design.
I believe the problem is not in https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Ui/view/base/web/js/lib/knockout/bindings/i18n.js but in the collection of the strings in the backend-code.

**Comment by engcom-Lima** [comment]
Hi @amenk,

I tried to reproduce this issue but did not find any problems in translating ‘foo bar 123’ into any language that I wish.

Here is what I did:

1. I created a new module to test it out.
2. I kept a string "foo bar 123" in the .phtml file. As per documentation kept it like __('Testing - foo bar 123') so that it can be translated properly. I have not added translation for 'Testing - ' so it is displayed in English only.
3. I sat local language to French (France) in Admin configurations.
4. I added French translation for it in Admin Panel like “barre de foo 123”
5. Flushed cache and checked on frontend. Result came with expected translations.

Please find the attached screenshot of the same. String is getting translated as expected:

<img width="1360" alt="Screenshot 2021-09-13 at 1 26 18 PM" src="https://user-images.githubusercontent.com/60198022/133046304-5e908bd7-5493-476c-89d2-e2b8c3d65d90.png">

Can you please try at your end and let me know if you are still facing this issue or update what you did differently at your end to get this issue ?

**Comment by mrtuvn** [comment]
@engcom-Lima you have to translate in file phtml use knockout syntax instead. That is correct case author try to mention here

**Comment by amenk** [comment]
Yes, normal .phtml translation on PHP level with the `__()` function of course works, knockout translation in `.html` also works. The problem is about the knockout translation in `.phtml` files.

**Comment by engcom-Lima** [comment]
Hi @amenk, @mrtuvn,

Thank you for the clarification.

Verified the issue and issue is reproducible on 2.4-develop branch.

I added string `<!-- ko i18n: 'foo bar 123' --><!-- /ko -->` in the same module which I created earlier and in same `.phtml` file but the translation did not worked with ko syntax. 

Language Translation is working with php tag and with knockout translation with `.html` also but not working with ko syntax. Based on it, confirming this issue.

Thanks

**Comment by mikimpe** [comment]
It appears that this issue affects also the translation of strings in the js section of a phtml file using the 'mage/translate' feature.
In this case, as a workaround, you can translate the string using PHP and then pass it to the js, for example like this:

```
let translatedString = "<?php echo __('Translate this string'); ?>";
```
